assembl/scripts/ini_files.py

```diff
@@ -144,72 +166,392 @@ def main():
-        })
-    else:
-        # dummy values
-        vars.update({
-            'VIRTUOSO_SERVER_PORT': 8890,
-            'VIRTUOSO_HOSTNAME': 'localhost',
-            'VIRTUOSO_PORT': 1111,
-            'VIRTUOSO_ROOT': '/usr',
-            'VIRTUOSO_ROOT_VAR': '/usr/var',
-            'VIRTUOSO_ROOT_LIB': '/usr/var/lib',
-            'VIRTUOSO_SUBDIR_NAME': '/usr/share/virtuoso/vad',
-        })
-
-    print vars
-    for fname in ('var/db/virtuoso.ini', 'odbc.ini', 'supervisord.conf',):
+
+    for fname in ('supervisord.conf',):
         print fname
-        tmpl = open(fname+'.tmpl').read()
-        inifile = open(fname, 'w')
-        inifile.write(tmpl % vars)
-        inifile.close()
+        with open(fname + '.tmpl') as tmpl, open(fname, 'w') as inifile:
+            inifile.write(tmpl.read() % vars)
+
+
+def combine_ini(config, overlay, adding=True):
+    """Overlay values of the second ini file on the first.
+
+    Returns a ConfigParser object
+    If adding is false, only existing values will be replaced.
+    """
+    config = asParser(config)
+    overlay = asParser(overlay)
+    for section in overlay.sections():
+        # Avoid including DEFAULTSECT
+        for key in overlay._sections[section]:
+            value = overlay.get(section, key)
+            if adding or config.has_option(section, key):
+                ensureSection(config, section)
+                config.set(section, key, value)
+    if overlay._defaults:
+        config._defaults.update(overlay._defaults)
+    return config
+
+
+def migrate_beaker_config(random_ini, overlay):
+    """Migrate old-style session... to beaker.session..."""
+    for section in random_ini.sections():
+        # Avoid including DEFAULTSECT
+        if not overlay.has_section(section):
+            continue
+        o_section = overlay._sections[section]
+        for key in random_ini._sections[section]:
+            if key.startswith("beaker."):
+                old_key = key[7:]
+                if old_key in o_section and key not in o_section:
+                    overlay.set(section, key, overlay.get(section, old_key))
+            elif key == 'session.secret':
+                alt_key = 'beaker.' + key
+                if alt_key in o_section and key not in o_section:
+                    overlay.set(section, key, overlay.get(section, alt_key))
+    return overlay
+
+
+def dump(ini_file):
+    """Dump the ini file, showing interpolations and errors."""
+    ini_file = asParser(ini_file, SafeConfigParser)
+    ini_file._defaults['here'] = os.getcwd()
+    for section in ini_file.sections():
+        print "\n[%s]\n" % section
+        for option in ini_file._sections[section]:
+            print option, '=',
+            try:
+                print ini_file.get(section, option)
+            except Exception as e:
+                print ini_file.get(section, option, True)
+                print "*"*15, "Could not interpolate"
+                print e
```

Comment:
print goes to stdout (unless you mess with sys.stdout.)
logs are logs. This subcommand is meant to only be used on CLI, not through fabric.